base/steer/FairRootManager.cxx

```diff
@@ -139,8 +194,109 @@ FairRootManager::~FairRootManager()
   LOG(DEBUG) << "Leave Destructor of FairRootManager" << FairLogger::endl;
   delete fEventHeader;
   delete fSourceChain;
+
+  // Global cleanup
+  --fgCounter;
+  if ( ! fgCounter ) {
+    delete fgIsFillLocks;
+    fgIsFillLocks = 0;
+  }
+
+  //
+#if ( ROOT_VERSION_CODE >= ROOT_VERSION(6,9,3) )
+  lk.unlock();
+#endif
+  if ( fDebugMT ) {
+    LOG(INFO) << "Released lock and done FairRootManager::~FairRootManager in "
+      << fId << " " << this << FairLogger::endl;
+  }
 }
+
+//
+// private methods for MT
+//
+
 //_____________________________________________________________________________
+void  FairRootManager::LogMessage(const TString& message)
+{
+/// Output one line log message followed with the instance Id and pointer
+
+  if ( fDebugMT ) {
+    LOG(INFO) << message << " in " << fId << " " << this << FairLogger::endl;
+  }
+}
+
+//_____________________________________________________________________________
+void  FairRootManager::FillWithTmpLock()
+{
+/// Fill the Root tree.
+
+  LogMessage("Going to lock for Fill");
+#if ( ROOT_VERSION_CODE >= ROOT_VERSION(6,9,3) )
+  TMCAutoLock lk(&tmpFillMutex2);
+#endif
+
+  LogMessage("Fill");
+  FillImpl();
+  LogMessage("Done Fill");
+
+  if ( fgIsFillLock ) {
+    // the access to TFile and TTree needs to be locked only until
+    // __after__ the first Fill
+    (*fgIsFillLocks)[fId] = false;
+    Bool_t isDoneAll = true;
+    Int_t counter = 1;
+    while ( isDoneAll && counter < fgCounter ) {
+      if ( (*fgIsFillLocks)[counter++] ) {
```

Comment:
I would like to avoid the post-increment operator here (it creates a copy of counter and it is harder to read). Could you please change it to something like:

```cpp
  if ( (*fgIsFillLocks)[counter] ) {
    isDoneAll = false;
    break;
  }
  ++counter;

Reply:
Ok, it is changed.